client/client.go

```
@@ -87,96 +101,543 @@ func (c *Client) Close() error {
+// it will first try to get the metadata associated with key from the Metadata servers
+func (c *Client) Read(key []byte) ([]byte, error) {
+	md, err := c.metaCli.Get(string(key))
+	if err != nil {
+		return nil, err
 	}
 
-	readPipe, err := pipe.NewReadPipe(c.conf, c.iyoToken)
+	w := &bytes.Buffer{}
+	err = c.readFWithMeta(md, w)
 	if err != nil {
 		return nil, err
 	}
 
+	return w.Bytes(), nil
+}
+
+// ReadF similar as Read but write the data to w instead of returning a slice of bytes
+func (c *Client) ReadF(key []byte, w io.Writer) error {
 	md, err := c.metaCli.Get(string(key))
 	if err != nil {
+		return err
+	}
+	return c.readFWithMeta(md, w)
+
+}
+
+// ReadWithMeta reads the value described by md
+func (c *Client) ReadWithMeta(md *meta.Meta) ([]byte, error) {
+	w := &bytes.Buffer{}
+	err := c.readFWithMeta(md, w)
+	if err != nil {
 		return nil, err
 	}
 
-	rp0 := readPipe.Readers[len(readPipe.Readers)-1]
-	chunkReader := rp0.(*chunker.BlockReader)
-	return nil, chunkReader.Restore(key, w, readPipe, md)
+	return w.Bytes(), nil
+}
+
+func (c *Client) readFWithMeta(md *meta.Meta, w io.Writer) error {
+	var (
+		aesgm encrypt.EncrypterDecrypter
+		block []byte
+		err   error
+	)
+
+	if c.policy.Encrypt {
+		aesgm, err = encrypt.NewEncrypterDecrypter(encrypt.Config{PrivKey: c.policy.EncryptKey, Type: encrypt.TypeAESGCM})
+		if err != nil {
+			return err
+		}
+	}
 
+	for _, chunk := range md.Chunks {
+
+		if c.policy.ReplicationMaxSize > 0 && chunk.Size <= uint64(c.policy.ReplicationMaxSize) {
+			block, err = c.replicateRead(chunk.Key, chunk.Shards)
+			if err != nil {
+				return err
+			}
+
+		} else if c.policy.DistributionNr > 0 && c.policy.DistributionRedundancy > 0 {
+			block, err = c.distributeRead(chunk.Key, int(chunk.Size), chunk.Shards)
+			if err != nil {
+				return err
+			}
+		} else {
+			if len(chunk.Shards) <= 0 {
+				panic("metadata corrupted, can't have a chunk without shard")
```

why `panic` instead of returning error?

I think it won't play well with our upcoming daemon